Turning circuits literate

Shdorsh opened the pull request:
Adds A text replacer circuit.
Who knows, maybe one day we will be able to have a real discussion with a circuit?

:cl: Shdorsh
add: Text replacer circuit
/:cl:

[why]: # While treatment of text seems like a very basic thing to do, current circuits don't allow to do this. As such, I added 4 new circuits to give much more possibilities when it comes to making a circuit create its own text.

Edit: regex got removed for security reasons and there was another find circuit already, so replace is the only one being added by this PR.

Shdorsh committed:
Create text.dm

Adds 4 text-treating circuits. Who knows, maybe one day we will be able to have a real discussion with a circuit?

Hate9 commented:
Y'know, with this and the min/max circuits, we're getting closer and closer to my dream of replacing the station AI with an actual AI made with circuits.

Shdorsh commented:
Soon, bro, soon.

Wilchenx commented:
You didn't change the name of your pr

nicbn commented:
Regex can crash the server depending on the needle - I was able to crash localhost sometimes while testing a paper markdown implementation with regex, but I'm not sure what tokens did it.

kevinz000 commented:
worried about the regex

ninjanomnom commented on code/modules/integrated_electronics/subtypes/text.dm:
```diff
@@ -0,0 +1,129 @@
+	)
+
+/obj/item/integrated_circuit/text/regex_finder/do_work()
+	var/regex/needle=regex(get_pin_data(IC_INPUT, 2))
```
Sorry but no, we can't have player determined regex run on the server

ninjanomnom commented on code/modules/integrated_electronics/subtypes/text.dm:
```diff
@@ -0,0 +1,129 @@
+	)
+
+/obj/item/integrated_circuit/text/regex_replacer/do_work()
+	var/regex/needle=regex(get_pin_data(IC_INPUT, 2))
```
same as above

ExcessiveUseOfCobblestone commented:
You put a lot of trust in your coding ability to have this webbited.

Shdorsh commented:
All the stuff I post has been tested on my homeserver, so that's why I put some trust behind it. I also didn't encounter problems with regex crashing the server and didn't know it could do that. Anyway, I'll just remove those 2 circuits from the PR.

Shdorsh committed:
Removing the regex stuff out of security reasons

Didn't know regex could crash, so here we go. I'll make sure to post another circuit to make up for that. What do you guys think about MMI controlled bots?

PKPenguin321 commented on code/modules/integrated_electronics/subtypes/text.dm:
```diff
@@ -0,0 +1,64 @@
+
+
+// - Text Finder - //
+/obj/item/integrated_circuit/text/text_finder
```
There is already a find text circuit

PKPenguin321 commented on code/modules/integrated_electronics/subtypes/text.dm:
```diff
@@ -0,0 +1,64 @@
+
+// - Text Replacer - //
+/obj/item/integrated_circuit/text/text_replacer
+	name = "text replacing circuit"
```
find-replace circuit

PKPenguin321 commented on code/modules/integrated_electronics/subtypes/text.dm:
```diff
@@ -0,0 +1,64 @@
+// - Text Replacer - //
+/obj/item/integrated_circuit/text/text_replacer
+	name = "text replacing circuit"
+	desc = "Replaces all of one bit of text with another"
```
Give it an input string (haystack), a string to search for (needle), and something to replace the needle with. It will output a modified version of the original string if it finds the needle.

PKPenguin321 reviewed (changes requested):
Points out seemingly redundant addition, grammar improvements

Shdorsh committed:
Fused find and replace text together

As you said, we could easily put both together, but it still needs to put out the position of the needle and whether it was found.

Shdorsh committed:
corrected some output pin issues

Shdorsh committed:
setting the right activators

this circuit now has the 2 on found and on not found activators

Shdorsh committed:
simplification of circuit

there we go, deleted some redundancies, due to other text finder circuit.

SpaceManiac reviewed (approved):
Code looks fine to me

nicbn commented:
Remove regex from changelo
